File: src/main/java/raccoonlang/parser/ast/InterfaceDefinitionAstNode.java
package raccoonlang.parser.ast;

import raccoonlang.parser.Parser;
import raccoonlang.tokenizer.Token;
import raccoonlang.tokenizer.TokenType;

public class InterfaceDefinitionAstNode {

    public ModifiersAstNode modifiersAstNode;
    public Token name;
    public GenericTypesAstNode genericTypesAstNode;

    public static InterfaceDefinitionAstNode tryParse(Parser parser) {
        Parser.ParserState parserState = parser.shelfState();
        InterfaceDefinitionAstNode interfaceDefinitionAstNode = new InterfaceDefinitionAstNode();

        interfaceDefinitionAstNode.modifiersAstNode = ModifiersAstNode.parse(parser);

        if(parser.peek().getTokenType() != TokenType.INTERFACE) {
            parser.restore(parserState);
            return null;
        }
        parser.take();

        interfaceDefinitionAstNode.name = parser.take(TokenType.IDENTIFIER);
        interfaceDefinitionAstNode.genericTypesAstNode = GenericTypesAstNode.tryParse(parser);



        return interfaceDefinitionAstNode;
    }
}

File: Parsing/AST/InterfaceDefinitionAstNode.cs
namespace Raccoonlang.Parsing.AST;

using LLVMSharp;

using Tokenizing;
using Utils;

public class InterfaceDefinitionAstNode : ITypeDef
{
    public ModifiersAstNode? Modifiers { get; set; }
    public Token? Name { get; set; }
    public GenericTypesAstNode? GenericTypes { get; set; }
    public GenericTypeConstraintsAstNode? GenericConstraints { get; set; }
    public InterfaceBodyAstNode? BodyContainer { get; set; }

    public static InterfaceDefinitionAstNode? TryParse(Parser parser)
    {
        InterfaceDefinitionAstNode node = new InterfaceDefinitionAstNode();

        var parserState = parser.ShelfState();
        node.Modifiers = ModifiersAstNode.Parse(parser);

        if (parser.Peek().Type != TokenType.Interface)
        {
            parser.RestoreState(parserState);
            return null;
        }

        parser.Take(TokenType.Interface);

        node.Name = parser.Take(TokenType.Identifier);
        node.GenericTypes = GenericTypesAstNode.TryParse(parser);
        node.GenericConstraints = GenericTypeConstraintsAstNode.Parse(parser);
        node.BodyContainer = InterfaceBodyAstNode.Parse(parser);

        return node;
    }

    public override string ToString() =>
        $"InterfaceDefinitionAstNode{{\nModifiers={Modifiers}, Name={Name}, GenericTypes={GenericTypes}, GenericConstraints={GenericConstraints}, BodyContainer={BodyContainer}}}";

    public void Compile(LLVMContextRef context, IRBuilder irBuilder, Module module)
    {
        CompilerLogger.Error("Not implemented yet!");
        Environment.Exit(1);
    }
}